File: Abeyonix_Frontend/src/api/companySettings.ts
import api from "./api";

import { CompanySettingsResponse, CompanySettingsPayload } from "../types/companySettings";



// =====================================================
// CREATE
// =====================================================

export const createCompanySettings = async (
  payload: CompanySettingsPayload
): Promise<CompanySettingsResponse> => {
  try {
    const formData = new FormData();

    formData.append("company_name", payload.company_name);

    if (payload.company_email)
      formData.append("company_email", payload.company_email);

    if (payload.company_phone)
      formData.append("company_phone", payload.company_phone);

    if (payload.address_line1)
      formData.append("address_line1", payload.address_line1);

    if (payload.city)
      formData.append("city", payload.city);

    if (payload.state)
      formData.append("state", payload.state);

    if (payload.postal_code)
      formData.append("postal_code", payload.postal_code);

    if (payload.country)
      formData.append("country", payload.country);

    if (payload.gst_number)
      formData.append("gst_number", payload.gst_number);

    if (payload.website)
      formData.append("website", payload.website);

    if (payload.logo) formData.append("logo", payload.logo);

    const response = await api.post<CompanySettingsResponse>(
      "/company-settings/",
      formData,
      {
        headers: {
          "Content-Type": "multipart/form-data",
        },
      }
    );

    return response.data;
  } catch (error: any) {
    throw (
      error?.response?.data?.detail ||
      "Failed to create company settings"
    );
  }
};

// =====================================================
// GET ALL
// =====================================================

export const getCompanySettings = async (): Promise<
  CompanySettingsResponse[]
> => {
  try {
    const response = await api.get<CompanySettingsResponse[]>(
      "/company-settings/"
    );

    return response.data;
  } catch (error: any) {
    throw (
      error?.response?.data?.detail ||
      "Failed to fetch company settings"
    );
  }
};

// =====================================================
// GET BY ID
// =====================================================

export const getCompanySettingsById = async (
  id: number
): Promise<CompanySettingsResponse> => {
  try {
    const response = await api.get<CompanySettingsResponse>(
      `/company-settings/${id}`
    );

    return response.data;
  } catch (error: any) {
    throw (
      error?.response?.data?.detail ||
      "Failed to fetch company settings"
    );
  }
};

// =====================================================
// UPDATE
// =====================================================

export const updateCompanySettings = async (
  id: number,
  payload: Partial<CompanySettingsPayload>
): Promise<CompanySettingsResponse> => {
  try {
    const formData = new FormData();

    if (payload.company_name)
      formData.append("company_name", payload.company_name);

    if (payload.company_email)
      formData.append("company_email", payload.company_email);

    if (payload.company_phone)
      formData.append("company_phone", payload.company_phone);

    if (payload.address_line1)
      formData.append("address_line1", payload.address_line1);

    if (payload.city)
      formData.append("city", payload.city);

    if (payload.state)
      formData.append("state", payload.state);

    if (payload.postal_code)
      formData.append("postal_code", payload.postal_code);

    if (payload.country)
      formData.append("country", payload.country);

    if (payload.gst_number)
      formData.append("gst_number", payload.gst_number);

    if (payload.website)
      formData.append("website", payload.website);

    if (payload.logo)
      formData.append("logo", payload.logo);

    const response = await api.put<CompanySettingsResponse>(
      `/company-settings/${id}`,
      formData,
      {
        headers: {
          "Content-Type": "multipart/form-data",
        },
      }
    );

    return response.data;
  } catch (error: any) {
    throw (
      error?.response?.data?.detail ||
      "Failed to update company settings"
    );
  }
};

// =====================================================
// DELETE
// =====================================================

export const deleteCompanySettings = async (
  id: number
): Promise<string> => {
  try {
    const response = await api.delete<{ message: string }>(
      `/company-settings/${id}`
    );

    return response.data.message;
  } catch (error: any) {
    throw (
      error?.response?.data?.detail ||
      "Failed to delete company settings"
    );
  }
};

File: Abeyonix_Backend/app/api/v1/company_settings.py
from fastapi import (
    APIRouter,
    Depends,
    HTTPException,
    status,
    UploadFile,
    File,
    Form
)

from sqlalchemy.orm import Session
from typing import Optional

from app.db.base import get_db
from app.models.company_settings import CompanySettings
from app.schemas.company_settings import CompanySettingsResponse

from app.utils.media import save_image, delete_image


router = APIRouter(
    prefix="/api/v1/company-settings",
    tags=["Company Settings"]
)


# =========================================================
# CREATE COMPANY SETTINGS
# =========================================================
@router.post(
    "/",
    response_model=CompanySettingsResponse,
    status_code=status.HTTP_201_CREATED
)
def create_company_settings(
    company_name: str = Form(...),
    company_email: Optional[str] = Form(None),
    company_phone: Optional[str] = Form(None),

    address_line1: Optional[str] = Form(None),
    city: Optional[str] = Form(None),
    state: Optional[str] = Form(None),
    postal_code: Optional[str] = Form(None),
    country: Optional[str] = Form(None),

    gst_number: Optional[str] = Form(None),
    website: Optional[str] = Form(None),

    logo: Optional[UploadFile] = File(None),

    db: Session = Depends(get_db),
):

    # Optional:
    # Only allow one company settings record
    existing = db.query(CompanySettings).first()
    if existing:
        raise HTTPException(
            status_code=400,
            detail="Company settings already exists"
        )

    logo_path = save_image(logo, "company") if logo else None

    company = CompanySettings(
        company_name=company_name,
        company_email=company_email,
        company_phone=company_phone,

        address_line1=address_line1,
        city=city,
        state=state,
        postal_code=postal_code,
        country=country,

        gst_number=gst_number,
        website=website,

        logo_url=logo_path
    )

    db.add(company)
    db.commit()
    db.refresh(company)

    return company


# =========================================================
# GET ALL COMPANY SETTINGS
# =========================================================
@router.get(
    "/",
    response_model=list[CompanySettingsResponse]
)
def get_company_settings(
    db: Session = Depends(get_db)
):
    return db.query(CompanySettings).all()


# =========================================================
# GET SINGLE COMPANY SETTINGS
# =========================================================
@router.get(
    "/{company_id}",
    response_model=CompanySettingsResponse
)
def get_company_settings_by_id(
    company_id: int,
    db: Session = Depends(get_db)
):
    company = (
        db.query(CompanySettings)
        .filter(CompanySettings.id == company_id)
        .first()
    )

    if not company:
        raise HTTPException(
            status_code=404,
            detail="Company settings not found"
        )

    return company


# =========================================================
# UPDATE COMPANY SETTINGS
# =========================================================
@router.put(
    "/{company_id}",
    response_model=CompanySettingsResponse
)
def update_company_settings(
    company_id: int,

    company_name: Optional[str] = Form(None),
    company_email: Optional[str] = Form(None),
    company_phone: Optional[str] = Form(None),

    address_line1: Optional[str] = Form(None),
    city: Optional[str] = Form(None),
    state: Optional[str] = Form(None),
    postal_code: Optional[str] = Form(None),
    country: Optional[str] = Form(None),

    gst_number: Optional[str] = Form(None),
    website: Optional[str] = Form(None),

    logo: Optional[UploadFile] = File(None),

    db: Session = Depends(get_db),
):

    company = (
        db.query(CompanySettings)
        .filter(CompanySettings.id == company_id)
        .first()
    )

    if not company:
        raise HTTPException(
            status_code=404,
            detail="Company settings not found"
        )

    # =====================================================
    # UPDATE FIELDS
    # =====================================================

    if company_name is not None:
        company.company_name = company_name

    if company_email is not None:
        company.company_email = company_email

    if company_phone is not None:
        company.company_phone = company_phone

    if address_line1 is not None:
        company.address_line1 = address_line1

    if city is not None:
        company.city = city

    if state is not None:
        company.state = state

    if postal_code is not None:
        company.postal_code = postal_code

    if country is not None:
        company.country = country

    if gst_number is not None:
        company.gst_number = gst_number

    if website is not None:
        company.website = website

    # =====================================================
    # UPDATE LOGO
    # =====================================================

    if logo:

        # delete old logo
        if company.logo_url:
            delete_image(company.logo_url)

        # save new logo
        company.logo_url = save_image(logo, "company")

    db.commit()
    db.refresh(company)

    return company


# =========================================================
# DELETE COMPANY SETTINGS
# =========================================================
@router.delete("/{company_id}")
def delete_company_settings(
    company_id: int,
    db: Session = Depends(get_db)
):

    company = (
        db.query(CompanySettings)
        .filter(CompanySettings.id == company_id)
        .first()
    )

    if not company:
        raise HTTPException(
            status_code=404,
            detail="Company settings not found"
        )

    # delete logo
    if company.logo_url:
        delete_image(company.logo_url)

    db.delete(company)
    db.commit()

    return {
        "message": "Company settings deleted successfully"
    }